### PR TITLE
Fix memory leak from static conditional weak tables used for analyzers

### PR DESCRIPTION
**Pull request by mavasani**
We use a bunch of static conditional weak tables in the repo to allow sharing expensive dataflow analysis computation results between analyzers. Most of these are keyed on Compilation, with the assumption that the entries will immediately get GC'ed when Compilation goes out of scope. However, the values stored in the cache have indirect cyclic reference to the Compilation, preventing GC from eagerly collecting them leading to extremely high memory usage and OOMs.
This change addresses this by introducing a new BoundedCache utility that mimics a bounded conditional weak table with a fixed size array of weak references to key-value pair entries. I verified that the original repro steps provided in https://github.com/dotnet/roslyn/issues/38330 lead to high memory usage and memory leak, and these no longer show up after the fix. Additioanlly, I also verified that we get reasonable amount of cache hits for the bounded caches, though we can fine tune the bound in future as required.

Fixes https://github.com/dotnet/roslyn/issues/38330. We will need to port the FlowAnalysis utilities package built on top of this change to Roslyn repo to avoid similar memory leaks from IDE Dispose dataflow analyzers.

### CODE REVIEW

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataConfig.cs [hunk 1]
```diff
@@ -57,92 +56,88 @@ internal class TaintedDataConfig
...
         }
 
-        private TaintedDataConfig(Compilation compilation)
+        private static TaintedDataConfig Create(Compilation compilation)
         {
-            this.WellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(compilation);
-            this.SourceSymbolMap = new Dictionary<SinkKind, Lazy<TaintedDataSymbolMap<SourceInfo>>>();
-            this.SanitizerSymbolMap = new Dictionary<SinkKind, Lazy<TaintedDataSymbolMap<SanitizerInfo>>>();
-            this.SinkSymbolMap = new Dictionary<SinkKind, Lazy<TaintedDataSymbolMap<SinkInfo>>>();
+            var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(compilation);
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataConfig.cs]: Sounds reasonable, can revert the usage of var :)

**Comment by weltkante** [comment]
@mavasani the motivation of this change is very unclear, how can a CWT prevent GC of the key? The whole (and only) point of CWT is to not prevent GC when the value references the key, otherwise you could use ordinary dictionaries with weak references.

The CWT is a collection of [Ephemerons](https://en.wikipedia.org/wiki/Ephemeron) which are conditional weak references keeping the value alive as long as the key is alive. So the value referencing the key does not matter because nobody should be having a reference to the value (the CWT value reference is ignored by the GC for purpose of determining liveness). If someone else is keeping a reference to the value (and thus indirectly to the key) they are the fault of the memory leak and won't go away if you replace the CWT.

We are relying on CWT working properly in our applications and if there is a bug which prevents keys from being collected when values reference the key then it should be fixed for everyone.

**Comment by mavasani** [comment]
@weltkante The primary issue here is not the usage of CWT, but the fact that we have multiple different CWTs with cycles. For example, if the values of the first CWT has a strong reference to the keys of another CWT, and the values of the second CWT has a strong reference to the keys of the first CWT. In our case, we had such strong reference cycles of between three or more CWTs, causing the leak.

Using a CWT is completely fine as long as you don't have such cycles - in fact we still have few CWTs in our repo where the value is purely data that cannot lead to cycles and it gets GC'ed just fine.

**Comment by weltkante** [comment]
@mavasani Thanks for responding but unfortunately your explanation does not make much sense. Looking at the implementation CWTs do not matter as far as the GC is concerned, how many you have and how their entries reference each others keys is irrelevant, each contained pair is an individual item to the GC and it doesn't know in which CWT it is stored. So any bug you can create with multiple CWTs should also exist in a single CWT.

Anyways, regardless of that distinction we also are using plenty of distinct CWTs in our products and we expect them to work with cycles present (for example to implement weak event subscriptions on arbitrary delegates so cycles are actually pretty common). If there exists a bug where the GC cannot track and resolve cycles between multiple entries under certain conditions than this needs to be reported and fixed. The whole point of introducing CWT was to allow the GC to solve this exact problem.

**Pull request merged**